core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala

```diff
@@ -4485,6 +4485,25 @@ class ReplicaManagerTest {
       assertTrue(response.usableBytes >= 0)
     }
   }
+
+  @Test
+  def checkRemoveMetricsCountMatchRegisterCount(): Unit = {
```

Comment:
I have an idea with which we can avoid changing the KafkaMetricsGroup. Could you please consider the following.

You can use mockito's `MockConstruction` to mock the KafkaMetricsGroup and count the number of invocations on that mock and later assert on number of invocations of add and remove. As an example, this test could be written as follows (rough example with Java code):

```
var numAddCount = 0;
var numRemoveCount = 0;
try (MockedConstruction<KafkaMetricsGroup> mockMetricsGroup = mockConstruction(KafkaMetricsGroup.class,
            (mock, context) -> {
                   doAnswer(invocation -> {
                        numAddCount++;
                    }).when(mock).newGauge(anyString());
                    
                    // similarly add mocks for newMeter etc.
                    
                    doAnswer(invocation -> {
                        numRemoveCount++;
                    }).when(mock).removeMetric(anyString());
        })) {
   val rm = new ReplicaManager(
      metrics = metrics,
      config = config,
      time = time,
      scheduler = new MockScheduler(time),
      logManager = mockLogMgr,
      quotaManagers = quotaManager,
      metadataCache = MetadataCache.zkMetadataCache(config.brokerId, config.interBrokerProtocolVersion),
      logDirFailureChannel = new LogDirFailureChannel(config.logDirs.size),
      alterPartitionManager = alterPartitionManager,
      threadNamePrefix = Option(this.getClass.getName))
      
   rm.shutdown()
   assertEquals(numAddCount, numRemoveCount)
 }
```

Reply:
Hello, I changed the KafkaMetricsGroup to facilitate the use of this kind of metric check for each class, and avoid mockConstruction(like above logic: numAddCount++, numRemoveCount, then compare) once for each class.
Do you think this is unnecessary? @divijvaidya

Reply:
I think that we can avoid changing the  `KafkaMetricsGroup` completely here and that is a preferred approach in my opinion because:

1. Methods in `KafkaMetricsGroup` may not be called from same thread? In that case, then we will have to use AtomicInteger for storing `numAddCount`. This will synchronize multiple threads on same lock hence, adding another point of contention.

2. Do we anticipate counters inside `KafkaMetricsGroup` to be used for any other purpose? If not, then we are adding counters only for writing unit test which is avoidable.

3. We are adding a potential source of bugs where if a new method is added is to `KafkaMetricsGroup`, the author will have to remember to correctly modify the counter.

Reply:
> I have an idea with which we can avoid changing the KafkaMetricsGroup. Could you please consider the following.
> 
> You can use mockito's `MockConstruction` to mock the KafkaMetricsGroup and count the number of invocations on that mock and later assert on number of invocations of add and remove. As an example, this test could be written as follows (rough example with Java code):
> 
> ```
> var numAddCount = 0;
> var numRemoveCount = 0;
> try (MockedConstruction<KafkaMetricsGroup> mockMetricsGroup = mockConstruction(KafkaMetricsGroup.class,
>             (mock, context) -> {
>                    doAnswer(invocation -> {
>                         numAddCount++;
>                     }).when(mock).newGauge(anyString());
>                     
>                     // similarly add mocks for newMeter etc.
>                     
>                     doAnswer(invocation -> {
>                         numRemoveCount++;
>                     }).when(mock).removeMetric(anyString());
>         })) {
>    val rm = new ReplicaManager(
>       metrics = metrics,
>       config = config,
>       time = time,
>       scheduler = new MockScheduler(time),
>       logManager = mockLogMgr,
>       quotaManagers = quotaManager,
>       metadataCache = MetadataCache.zkMetadataCache(config.brokerId, config.interBrokerProtocolVersion),
>       logDirFailureChannel = new LogDirFailureChannel(config.logDirs.size),
>       alterPartitionManager = alterPartitionManager,
>       threadNamePrefix = Option(this.getClass.getName))
>       
>    rm.shutdown()
>    assertEquals(numAddCount, numRemoveCount)
>  }
> ```

Sorry, I tried my best to write this unit test according to this method, but since newGauge has a return value, there is no way to use doAnswer. It seems that KafkaMetricsGroup must be changed? @divijvaidya